yarn-project/prover-client/src/prover/bb_prover.ts

```diff
@@ -0,0 +1,256 @@
+
+    const witnessMap = convertMergeRollupInputsToWitnessMap(input);
+
+    const [outputWitness, proof] = await this.createProof(witnessMap, 'MergeRollupArtifact');
+
+    const result = convertMergeRollupOutputsFromWitnessMap(outputWitness);
+
+    return Promise.resolve([result, proof]);
+  }
+
+  /**
+   * Simulates the root rollup circuit from its inputs.
+   * @param input - Inputs to the circuit.
+   * @returns The public inputs as outputs of the simulation.
+   */
+  public async getRootRollupProof(input: RootRollupInputs): Promise<[RootRollupPublicInputs, Proof]> {
+    // verify the inputs
+    await Promise.all(input.previousRollupData.map(prev => this.verifyPreviousRollupProof(prev)));
+
+    const witnessMap = convertRootRollupInputsToWitnessMap(input);
+
+    const [outputWitness, proof] = await this.createProof(witnessMap, 'RootRollupArtifact');
+
+    await this.verifyProof('RootRollupArtifact', proof);
+
+    const result = convertRootRollupOutputsFromWitnessMap(outputWitness);
+    return Promise.resolve([result, proof]);
+  }
+
+  private async init() {
+    const promises = [];
+    for (const circuitName in ServerCircuitArtifacts) {
+      const verificationKeyPromise = generateKeyForNoirCircuit(
+        this.config.bbBinaryPath,
+        this.config.bbWorkingDirectory,
+        circuitName,
+        ServerCircuitArtifacts[circuitName as ServerProtocolArtifact],
+        'vk',
+        logger.debug,
+      ).then(result => {
+        if (result.status == BB_RESULT.FAILURE) {
+          logger.error(`Failed to generate verification key for circuit ${circuitName}`);
+          return;
+        }
+        logger.info(`Generated verification key for circuit ${circuitName} at ${result.path!}`);
+        this.verificationKeyDirectories.set(circuitName as ServerProtocolArtifact, result.path!);
+      });
+      promises.push(verificationKeyPromise);
+    }
+    await Promise.all(promises);
+  }
+
+  private async createProof(witnessMap: WitnessMap, circuitType: ServerProtocolArtifact): Promise<[WitnessMap, Proof]> {
+    // Create random directory to be used for temp files
+    const bbWorkingDirectory = `${this.config.bbWorkingDirectory}/${randomBytes(8).toString('hex')}`;
+    await fs.mkdir(bbWorkingDirectory, { recursive: true });
+
+    // Have the ACVM write the partial witness here
+    const outputWitnessFile = `${bbWorkingDirectory}/partial-witness.gz`;
+
+    // Generate the partial witness using the ACVM
+    // A further temp directory will be created beneath ours and then cleaned up after the partial witness has been copied to our specified location
+    const simulator = new NativeACVMSimulator(
+      this.config.acvmWorkingDirectory,
+      this.config.acvmBinaryPath,
+      outputWitnessFile,
+    );
+
+    const artifact = ServerCircuitArtifacts[circuitType];
+
+    logger.debug(`Generating witness data for ${circuitType}`);
+
+    const outputWitness = await simulator.simulateCircuit(witnessMap, artifact);
```

Comment:
Is this saying that the actors in our pool parallel processors will be "proovers" as well as "simulators"? That makes me a bit nervous if we want to interface with semi-opaque proving pools.

Reply:
We kind of conflate 2 different things unfortunately. Simulation here is really witness generation, the first step of proving.

When we talk about simulation elsewhere we really mean the execution of public functions by the sequencer (or a searcher for example). That's different and won't be done by provers.